source/common/tls/client_context_impl.h

```diff
@@ -45,22 +45,27 @@ class ClientContextImpl : public ContextImpl, public Envoy::Ssl::ClientContext {
 
   absl::StatusOr<bssl::UniquePtr<SSL>>
   newSsl(const Network::TransportSocketOptionsConstSharedPtr& options,
-         Upstream::HostDescriptionConstSharedPtr host) override;
+         const Upstream::HostDescriptionConstSharedPtr& host) override;
 
 private:
   ClientContextImpl(Stats::Scope& scope, const Envoy::Ssl::ClientContextConfig& config,
                     Server::Configuration::CommonFactoryContext& factory_context,
                     absl::Status& creation_status);
 
-  int newSessionKey(SSL_SESSION* session);
+  int newSessionKey(SSL_SESSION* session, const Upstream::HostDescriptionConstSharedPtr& host);
+  static int sslSocketUpstreamHostIndex();
+  uint64_t sessionCacheKey(const Upstream::HostDescriptionConstSharedPtr& host);
 
   const std::string server_name_indication_;
   const bool auto_host_sni_;
   const bool allow_renegotiation_;
   const bool enforce_rsa_key_usage_;
   const size_t max_session_keys_;
-  absl::Mutex session_keys_mu_;
-  std::deque<bssl::UniquePtr<SSL_SESSION>> session_keys_ ABSL_GUARDED_BY(session_keys_mu_);
+  const absl::optional<const Ssl::ClientContextConfig::PerHostSessionCacheConfig>
+      per_host_session_cache_config_;
+  absl::Mutex session_keys_map_mu_;
+  absl::flat_hash_map<uint64_t, std::deque<bssl::UniquePtr<SSL_SESSION>>>
```

Comment:
You're manually hashing a string into an int64. I think it would be clearer to use a string as the key here instead.